src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -880,20 +904,29 @@ private static void ParseHeaderNameValue(Span<byte> line, HttpResponseMessage re
                 if (pos == line.Length)
                 {
                     // Invalid header line that doesn't contain ':'.
-                    ThrowInvalidHttpResponse();
+                    TryThrowInvalidHttpResponse(isFromTrailer);
+                    return;
                 }
             }
 
             if (pos == 0)
             {
                 // Invalid empty header name.
-                ThrowInvalidHttpResponse();
+                TryThrowInvalidHttpResponse(isFromTrailer);
+                return;
             }
 
             if (!HeaderDescriptor.TryGet(line.Slice(0, pos), out HeaderDescriptor descriptor))
             {
-                // Invalid header name
-                ThrowInvalidHttpResponse();
+                // Invalid header name. Need to throw.
+                TryThrowInvalidHttpResponse(false);
+            }
+
+            if (isFromTrailer && descriptor.KnownHeader != null && s_disallowedHeaders.Contains(descriptor.KnownHeader))
```

Comment:
I have to ask, is it really worthwhile to filter out these headers? I realize the RFC says "A recipient MUST ignore..." but that doesn't necessarily mean we can't just pass them to the user and expect them to ignore them. The "recipient" here is some combination of HttpClient and the user of HttpClient, and I think there's a reasonable argument to be made that the user of HttpClient is better suited to decide whether these are meaningful or not.

Reply:
I like that if for no other reason than it keeps the code simpler and cheaper.

Reply:
Protocol MUSTs are best handled by the HttpClient implementation, consumers of HttpClient generally aren't familiar with the protocol details. If they MUST be ignored then the consumer would need to make the same check we are, we might as well do it for them. These headers were banned because they pose either functional or security risks to the consumer. Putting trailers into a seperate collection significantly reduces the risks, but it would still be bad if something aggregated these headers with the others downstream.